mesonbuild/compilers/c.py

```diff
@@ -761,17 +761,27 @@ def get_library_naming(self, env, libtype):
     def find_library(self, libname, env, extra_dirs, libtype='default'):
         # These libraries are either built-in or invalid
         if libname in self.ignore_libs:
-            return []
+            return [], []
+
         # First try if we can just add the library as -l.
         code = 'int main(int argc, char **argv) { return 0; }'
         if extra_dirs and isinstance(extra_dirs, str):
             extra_dirs = [extra_dirs]
+
         # Gcc + co seem to prefer builtin lib dirs to -L dirs.
         # Only try to find std libs if no extra dirs specified.
         if not extra_dirs and libtype == 'default':
             args = ['-l' + libname]
             if self.links(code, env, extra_args=args):
-                return args
+                return args, []
+        # Some of the BSD's (freebsd and dragonfly) install packages into
+        # /usr/local, but that's not in the default search path. On those
+        # platforms check the ports dir and append the proper flags if it's
+        # found there.
+        if is_dragonflybsd():
+            args = ['-L/usr/local/lib', '-l' + libname]
+            if self.links(code, env, extra_args=args):
+                return args, ['-isystem/usr/local/include']
```

Comment:
Do we maybe want to always search first for libraries in /usr/local/lib on *bsd? The only case I can think of is when the same library is in both, and don't we want to prefer the ports version in that case?

Reply:
Also, shouldn't this be `if is_bsd():`?

Reply:
I think you're right. pkg-config checks /usr/local before /usr on dragonfly, so I think you're right and we should check /usr/local first too.

Reply:
and no, it should. Only freebsd and dragonfly do this, netbsd puts their packages in /usr/pkg, I'm not sure what openbsd does, haven't gotten there yet.

Reply:
Then it should be `if is_dragonflybsd() or is_freebsd():`?

Reply:
yeah, but we don't have an `is_freebsd` yet :)

I've almost got everything passing or skipping gracefully on dragonfly, and freebsd is my next target. There's some other awfulness there because they have an ancient version of GCC (4.2). I think it might be better to default to clang on more modern versions of FreeBSD. Dragonfly has 5.4, which is still old, but not prehistoric.

Reply:
Is it required for a build system to handle /usr/local? Since most ports set `CPPFLAGS=-I/usr/local/include` and `LDFLAGS=-L/usr/local/lib` in their Makefiles and users are expected to do the same thing themselves in their environments when not using ports, meson should be able to pick these required arguments from environment variables. If meson doesn't pass the value of `CPPFLAGS` or `LDFLAGS` to the compiler in some modules, I think we should consider it a bug and I hope we are not going to make a change that is likely to hide bugs. This kind of bug does exist in GNOME module.

Both /usr/local (on DragonflyBSD, FreeBSD, OpenBSD) and /usr/pkg (on NetBSD) are just recommended and default values of installation prefix. It can be changed by users and the ports framework should never hard-code the prefix. When a maintainer of a port finds a software hard-coding /usr/local on FreeBSD, it is likely to see the software being patched to respect the prefix chosen by the user.

FreeBSD 10 and later uses Clang as its default compiler, at least on x86. /usr/bin/cc is a hard link to clang and gcc is not installed by default. Do you run FreeBSD on a tier-2 or tier-3 platform where gcc is still the default? FreeBSD 11.1 has Clang 4.0 which is be good enough to build most packages.

I ran pkg-config under truss on FreeBSD, and yes, it checked /usr/local before /usr. I could not find any port installing a .pc file having the same name as files installed in /usr, so the .pc files search order of /usr/local and /usr is probably not important on FreeBSD.

Reply:
> Since most ports set CPPFLAGS=-I/usr/local/include and LDFLAGS=-L/usr/local/lib in their Makefiles and users are expected to do the same thing themselves in their environments when not using ports, meson should be able to pick these required arguments from environment variables.

We do this by default already.